### PR TITLE
113 include theoretical pump cost in tco

### PR DESCRIPTION
**Pull request by jimrojerTNO**
This PR adds the theoretical pump power to the computations and uses it to compute a pump cost based on the electricity price specified on a electricity carrier (for now we assume the user has one electricity carrier).

- [x] We need to utilize the electicity carrier info for the pumping cost
- [x] We need to add the cost to the variable operational cost variable
- [x] We need to explicitly add the pumps to the TCO goal

For now we have no test case as I want to extend on the test script of Kobus in head_loss. But you can run the HeatProblemHydraulic problem in source_sink_pipe to see that it works.

closes:
#123 
#122 
#118
#113

### CODE REVIEW

File: tests/models/source_pipe_sink/src/double_pipe_heat.py [hunk 1]
```diff
@@ -47,11 +54,42 @@ def __init__(self, *args, **kwargs):
...
+        self.heat_network_settings["minimize_head_losses"] = True
+
+    def heat_network_options(self):
+        options = super().heat_network_options()
+
+        return options
+
+    def solver_options(self):
+        options = super().solver_options()
+        options["solver"] = "gurobi"
```
Comment (KobusVanRooyen) [inline tests/models/source_pipe_sink/src/double_pipe_heat.py:81]: The "gurobi" option should be commented out

File: tests/models/unit_cases/case_1a/src/run_1a.py [hunk 2]
```diff
@@ -125,8 +125,9 @@ class QTHProblem(
 
 
 if __name__ == "__main__":
+    # sol = run_optimization_problem(HeatProblemTvar)
     sol = run_optimization_problem(
-        HeatProblemTvar,
+        HeatProblem,
```
Comment (KobusVanRooyen) [inline tests/models/unit_cases/case_1a/src/run_1a.py:130]: Should HeatProblem or HeatProblemTvar be used in the optim? Maybe deleted the commented out code if it is not needed

File: src/rtctools_heat_network/heat_physics_mixin.py [hunk 3]
```diff
@@ -2580,6 +2602,79 @@ def __heat_pump_cop_constraints(self, ensemble_member):
...
+            big_m = (
+                2.0
+                * self.bounds()[f"{b}.HeatIn.Q"][1]
+                * self.__maximum_total_head_loss
+                * 10.2
+                * 1.0e3
+            )
+            if self.heat_network_settings["head_loss_option"] != HeadLossOption.NO_HEADLOSS:
+
+                # During charging we want a minimum pressure drop like a demand
```
[the remark this answers is not in the document]
Comment (FJanssen-TNO) [reply src/rtctools_heat_network/heat_physics_mixin.py:2632]: I'm not sure if I entirely understand these constraints.

File: src/rtctools_heat_network/_darcy_weisbach.py [hunk 4]
```diff
@@ -1,41 +1,37 @@
...
-
-        from iapws import IAPWS95
-
-        import numpy as np
-
-        res = []
-        for t in np.linspace(20, 130, 1000):
-            res.append((t, IAPWS95(T=273.15 + t, P=0.5).nu))
-        print(np.polyfit(*zip(*res), 4))
+    The kinematic viscosity barely changes with pressure for water.
```
[the remark this answers is not in the document]
Comment (FJanssen-TNO) [reply src/rtctools_heat_network/_darcy_weisbach.py]: updated

File: src/rtctools_heat_network/_darcy_weisbach.py [hunk 5]
```diff
@@ -1,41 +1,37 @@
...
+        return cP.CoolProp.PropsSI("V", "T", 273.15 + temperature, "P", pressure, "HYDROGEN")
+    elif network_type == NetworkSettings.NETWORK_TYPE_GAS:
+        pressure = pressure if pressure else 101325
+        return cP.CoolProp.PropsSI(
+            "V",
+            "T",
+            273.15 + temperature,
+            "P",
+            pressure,
+            "methane[0.813]&ethane[0.0285]&propane[0.0037]&butane[0.0014]&n-Pentane[0.0004]"
```
[the remark this answers is not in the document]
Comment (FJanssen-TNO) [reply src/rtctools_heat_network/_darcy_weisbach.py]: updated

File: src/rtctools_heat_network/_darcy_weisbach.py [hunk 6]
```diff
@@ -1,41 +1,37 @@
...
+        return IAPWS95(T=273.15 + temperature, P=0.5).nu
+    elif network_type == NetworkSettings.NETWORK_TYPE_HYDROGEN:
+        pressure = pressure if pressure else 101325
+        return cP.CoolProp.PropsSI("V", "T", 273.15 + temperature, "P", pressure, "HYDROGEN")
+    elif network_type == NetworkSettings.NETWORK_TYPE_GAS:
+        pressure = pressure if pressure else 101325
+        return cP.CoolProp.PropsSI(
+            "V",
+            "T",
+            273.15 + temperature,
```
[the remark this answers is not in the document]
Comment (FJanssen-TNO) [reply src/rtctools_heat_network/_darcy_weisbach.py:33]: incorporated in the docstring as it is different for gas/hydrogen using coolprop and for water using IAPWS95

File: src/rtctools_heat_network/esdl/esdl_additional_vars_mixin.py [hunk 7]
```diff
@@ -19,6 +19,12 @@ def esdl_heat_model_options(self):
         options["min_fraction_tank_volume"] = 0.0
         return options
 
+    def temperature_carriers(self):
+        return self.esdl_carriers_typed(type="heat")
```
[the remark this answers is not in the document]
Comment (FJanssen-TNO) [reply src/rtctools_heat_network/esdl/esdl_additional_vars_mixin.py:23]: updated

File: src/rtctools_heat_network/esdl/esdl_heat_model.py [hunk 8]
```diff
@@ -85,6 +89,30 @@ def _rho_cp_modifiers(self) -> Dict:
         """
         return dict(rho=self.rho, cp=self.cp)
 
+    def get_density(self, asset_name, carrier):
+        temperature = 20.0
```
[the remark this answers is not in the document]
Comment (FJanssen-TNO) [reply src/rtctools_heat_network/esdl/esdl_heat_model.py:96]: updated

File: src/rtctools_heat_network/esdl/esdl_heat_model.py [hunk 9]
```diff
@@ -85,6 +89,30 @@ def _rho_cp_modifiers(self) -> Dict:
...
+
+        if NetworkSettings.NETWORK_TYPE_GAS in carrier.name:
+            density = cP.CoolProp.PropsSI(
+                "D",
+                "T",
+                273.15 + temperature,
+                "P",
+                carrier.pressure,
+                "methane[0.813]&ethane[0.0285]&propane[0.0037]&butane[0.0014]&n-Pentane[0.0004]"
```
[the remark this answers is not in the document]
Comment (FJanssen-TNO) [reply src/rtctools_heat_network/esdl/esdl_heat_model.py]: updated

File: src/rtctools_heat_network/esdl/esdl_heat_model.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (FJanssen-TNO) [reply src/rtctools_heat_network/esdl/esdl_heat_model.py:96]: updated

File: tests/models/unit_cases/case_1a/src/run_1a.py (hunk 2 above)
Comment (FJanssen-TNO) [reply tests/models/unit_cases/case_1a/src/run_1a.py:129]: Doesn't matter, depends on what you try to run locally.

File: src/rtctools_heat_network/electricity_physics_mixin.py [hunk 10]
```diff
@@ -52,6 +52,13 @@ def heat_network_options(self):
 
         return options
 
+    def electricity_carriers(self):
+        """
+        This function should be overwritten by the problem and should give a dict with the
+        carriers as keys and a list of temperatures as values.
```
[the remark this answers is not in the document]
Comment (jimrojerTNO) [reply src/rtctools_heat_network/electricity_physics_mixin.py:58]: yeah copying error

File: src/rtctools_heat_network/esdl/asset_to_component_base.py [hunk 11]
```diff
@@ -426,6 +427,35 @@ def _get_connected_q_nominal(self, asset: Asset) -> Union[float, Dict]:
                 raise _RetryLaterException(
                     f"Could not determine nominal discharge for {asset.asset_type} '{asset.name}'"
                 )
+        elif (
```
[the remark this answers is not in the document]
Comment (jimrojerTNO) [reply src/rtctools_heat_network/esdl/asset_to_component_base.py:430]: agree

File: src/rtctools_heat_network/esdl/esdl_mixin.py [hunk 12]
```diff
@@ -315,16 +315,34 @@ def esdl_assets(self) -> Dict[str, Asset]:
...
         property method to retrieve the esdl carriers which are a private attribute of the class.
 
         Returns
         -------
         A dict with the id of the carrier and the attributes in the value
         """
+
         return self._esdl_carriers
 
+    def esdl_carriers_typed(self, type=None) -> Dict[str, Dict[str, Any]]:
```
[the remark this answers is not in the document]
Comment (jimrojerTNO) [reply src/rtctools_heat_network/esdl/esdl_mixin.py:329]: not sure anymore, I think your suggestion is valid point I would prefer that we implement what you suggest

File: src/rtctools_heat_network/gas_physics_mixin.py [hunk 13]
```diff
@@ -103,7 +103,7 @@ def __init__(self, *args, **kwargs):
             "pipe_minimum_pressure": -np.inf,
             "pipe_maximum_pressure": np.inf,
         }
-        self._head_loss_class = HeadLossClass(self.gas_network_settings)
+        self._gn_head_loss_class = HeadLossClass(self.gas_network_settings)
```
[the remark this answers is not in the document]
Comment (jimrojerTNO) [reply src/rtctools_heat_network/gas_physics_mixin.py:106]: I definitely agree with the consistent naming between heat and gas, so self._hn_head_loss_class is a good idea.

personally I like the shorter names, but I am not against using the full names

File: src/rtctools_heat_network/heat_physics_mixin.py [hunk 14]
```diff
@@ -2580,6 +2602,79 @@ def __heat_pump_cop_constraints(self, ensemble_member):
 
         return constraints
 
+    def __storage_hydrualic_power_path_constraints(self, ensemble_member):
```
[the remark this answers is not in the document]
Comment (jimrojerTNO) [reply src/rtctools_heat_network/heat_physics_mixin.py]: fair point

File: src/rtctools_heat_network/heat_physics_mixin.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jimrojerTNO) [reply src/rtctools_heat_network/heat_physics_mixin.py:2637]: The idea was to have depending on the charging or discharging have the storage assets behave like a demand or source respectively also for the hydraulic power. Meaning a hydraulic power for a minimum pressuredrop during charging similar to the demands and a pump power like a source during discharging.

In my mind this was logical, but maybe I am missing something

[the remark this answers is not in the document]
Comment (KobusVanRooyen) [reply src/rtctools_heat_network/heat_physics_mixin.py:2641]: @femke please check my updated docstring. After discussion with Jim

File: src/rtctools_heat_network/esdl/asset_to_component_base.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (KobusVanRooyen) [reply src/rtctools_heat_network/esdl/asset_to_component_base.py:430]: Comment has been added

File: src/rtctools_heat_network/electricity_physics_mixin.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (KobusVanRooyen) [reply src/rtctools_heat_network/electricity_physics_mixin.py]: Changed temperature to voltage

File: src/rtctools_heat_network/gas_physics_mixin.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (KobusVanRooyen) [reply src/rtctools_heat_network/gas_physics_mixin.py:106]: I implemented hn_ -->> self._hn_head_loss_class in heat physics mixin

File: src/rtctools_heat_network/esdl/esdl_mixin.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (KobusVanRooyen) [reply src/rtctools_heat_network/esdl/esdl_mixin.py:329]: Left as is for now

File: src/rtctools_heat_network/heat_physics_mixin.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (KobusVanRooyen) [reply src/rtctools_heat_network/heat_physics_mixin.py:2641]: Checked

**Review by KobusVanRooyen — APPROVED** [review]

**Pull request merged**